### ISSUE TITLE
[address bug]Segment Fault crash in binary idlc

### ISSUE THREAD

**Issue by Yhcrown** [eclipse-cyclonedds/cyclonedds#1370]
**Hi, developers of cyclonedds:**
Recently, I tested the binary idlc instrumented with ASAN by our fuzzing tool. Unfortunately, it incured a segment fault crash. I recorded the error information with address sanitizer:

crash input:
https://github.com/Yhcrown/fuzzing-crashes/blob/main/validated/idlc/105
error output:
```
/workspace/crash/cyclonedds/check_build/bin/idlc/105:1: warning: End of input with no newline, supplemented newline
#1"="00000���11111111111111111:::::::::::::::::::::::;:::::::::::::::::::::9:::::::::::S::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::#1::::::::::#::::::::::::::::::z::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::111111
/workspace/crash/cyclonedds/check_build/bin/idlc/105:4:11: Extra tokens at end of line marker
AddressSanitizer:DEADLYSIGNAL
=================================================================
==68523==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x56495b574405 bp 0x7ffc0bcebcb0 sp 0x7ffc0bceb460 T0)
==68523==The signal is caused by a READ memory access.
==68523==Hint: address points to the zero page.
    #0 0x56495b574405 in strcmp (/workspace/test/cyclonedds/bin/idlc+0x5a405) (BuildId: e526cf43a1a1020257bfd1c9437645cac70cf01e)
    #1 0x7fcc380ee412 in push_file /workspace/test/cyclonedds/src/idl/src/directive.c:60:18
    #2 0x7fcc380ee412 in push_line /workspace/test/cyclonedds/src/idl/src/directive.c:177:16
    #3 0x7fcc380ec1b8 in idl_parse_directive /workspace/test/cyclonedds/src/idl/src/directive.c:604:12
    #4 0x7fcc380f716a in idl_parse /workspace/test/cyclonedds/src/idl/src/processor.c:529:15
    #5 0x56495b61c730 in idlc_parse /workspace/test/cyclonedds/src/tools/idlc/src/idlc.c:372:11
    #6 0x56495b61bc84 in main /workspace/test/cyclonedds/src/tools/idlc/src/idlc.c:703:16
    #7 0x7fcc379ecd8f  (/lib/x86_64-linux-gnu/libc.so.6+0x29d8f) (BuildId: 69389d485a9793dbe873f0ea2c93e02efaa9aa3d)
    #8 0x7fcc379ece3f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x29e3f) (BuildId: 69389d485a9793dbe873f0ea2c93e02efaa9aa3d)
    #9 0x56495b55d8e4 in _start (/workspace/test/cyclonedds/bin/idlc+0x438e4) (BuildId: e526cf43a1a1020257bfd1c9437645cac70cf01e)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV (/workspace/test/cyclonedds/bin/idlc+0x5a405) (BuildId: e526cf43a1a1020257bfd1c9437645cac70cf01e) in strcmp
==68523==ABORTING
```

### Verification steps：

```
git clone https://github.com/eclipse-cyclonedds/cyclonedds
cd cyclonedd
mkdir check_build && cd check_build
cmake ../ -DCMAKE_C_COMPILER=clang  -DCMAKE_CXX_COMPILER=clang++ -DCMAKE_C_FLAGS="-fsanitize=address -U_FORTIFY_SOURCE" -DCMAKE_CXX_FLAGS="-fsanitize=address -U_FORTIFY_SOURCE"
cmake --build .
cd bin
./idlc ./crash_input
```

### enviroment:
Ubuntu 22.04 LTS
Linux lab-pro 4.15.0-147-generic
gcc (Ubuntu 11.2.0-19ubuntu1) 11.2.0
Ubuntu clang version 14.0.6

Any help would be greatly appreciated from you :D
**Thanks & Best regards !**

**Comment by dpotman** [comment]
@Yhcrown thanks for reporting these issues! I've analyzed them using your input files, and fixes are in PR #1381 (the issues from #1368 and #1369 are also included in that PR). Please let us know if you find any additional issues!

**Comment by Yhcrown** [comment]
> @Yhcrown thanks for reporting these issues! I've analyzed them using your input files, and fixes are in PR #1381 (the issues from #1368 and #1369 are also included in that PR). Please let us know if you find any additional issues!

Thank you very much for fixing my issue, it is also an encouragement for our work. I found three other additional bugs in the last few days. one of them is of the same type as this issue, a segment fault. 
Input file.
https://github.com/Yhcrown/fuzzing-crashes/blob/main/validated/idlc/229
output:
```
AddressSanitizer:DEADLYSIGNAL
=================================================================
==33456==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000028 (pc 0x7fc1f77cd6f7 bp 0x7ffd3ee93310 sp 0x7ffd3ee931c0 T0)
==33456==The signal is caused by a READ memory access.
==33456==Hint: address points to the zero page.
    #0 0x7fc1f77cd6f7 in push_line /workspace/test/cyclonedds/src/idl/src/directive.c
    #1 0x7fc1f77cb1b8 in idl_parse_directive /workspace/test/cyclonedds/src/idl/src/directive.c:604:12
    #2 0x7fc1f77d616a in idl_parse /workspace/test/cyclonedds/src/idl/src/processor.c:529:15
    #3 0x55582976f730 in idlc_parse /workspace/test/cyclonedds/src/tools/idlc/src/idlc.c:372:11
    #4 0x55582976ec84 in main /workspace/test/cyclonedds/src/tools/idlc/src/idlc.c:703:16
    #5 0x7fc1f70cad8f  (/lib/x86_64-linux-gnu/libc.so.6+0x29d8f) (BuildId: 69389d485a9793dbe873f0ea2c93e02efaa9aa3d)
    #6 0x7fc1f70cae3f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x29e3f) (BuildId: 69389d485a9793dbe873f0ea2c93e02efaa9aa3d)
    #7 0x5558296b08e4 in _start (/workspace/test/cyclonedds/bin/idlc+0x438e4) (BuildId: e526cf43a1a1020257bfd1c9437645cac70cf01e)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV /workspace/test/cyclonedds/src/idl/src/directive.c in push_line
==33456==ABORTING

```

The other two types of crashes I will report in a new issue. Thanks again.

**Comment by dpotman** [comment]
The fix for the `229` issue is merged in #1384 , so @Yhcrown if you could confirm that the fix works, we don't have to re-open this issue

**Comment by Yhcrown** [comment]
After my testing, this bug has been well fixed.

> The fix for the `229` issue is merged in #1384 , so @Yhcrown if you could confirm that the fix works, we don't have to re-open this issue
